homeassistant/components/samsungtv/media_player.py

```diff
@@ -239,33 +244,66 @@ async def _async_startup_app_list(self) -> None:
             with contextlib.suppress(UpnpConnectionError):
-                self._upnp_device = await upnp_factory.async_create_device(
+                upnp_device = await upnp_factory.async_create_device(
                     self._ssdp_rendering_control_location
                 )
-
-    def _get_upnp_service(self, log: bool = False) -> UpnpService | None:
-        if self._upnp_device is None:
-            if log:
-                LOGGER.info("Upnp services are not available on %s", self._host)
-            return None
-
-        if service := self._upnp_device.services.get(UPNP_SVC_RENDERING_CONTROL):
-            return service
-
-        if log:
-            LOGGER.info(
-                "Upnp service %s is not available on %s",
-                UPNP_SVC_RENDERING_CONTROL,
-                self._host,
+            if not upnp_device:
+                return
+            _, event_ip = await async_get_local_ip(
+                self._ssdp_rendering_control_location, self.hass.loop
+            )
+            source = (event_ip or "0.0.0.0", 0)
+            self._upnp_server = AiohttpNotifyServer(
+                requester=upnp_requester,
+                source=source,
+                callback_url=None,
+                loop=self.hass.loop,
             )
-        return None
+            await self._upnp_server.async_start_server()
+            self._dmr_device = DmrDevice(upnp_device, self._upnp_server.event_handler)
+
+            try:
+                self._dmr_device.on_event = self._on_upnp_event
+                await self._dmr_device.async_subscribe_services(auto_resubscribe=True)
+            except UpnpResponseError as err:
+                # Device rejected subscription request. This is OK, variables
+                # will be polled instead.
+                LOGGER.debug("Device rejected subscription: %r", err)
+            except UpnpError as err:
+                # Don't leave the device half-constructed
+                self._dmr_device.on_event = None
+                self._dmr_device = None
+                await self._upnp_server.async_stop_server()
+                self._upnp_server = None
+                LOGGER.debug("Error while subscribing during device connect: %r", err)
+                raise
+
+    async def _async_shutdown_dmr(self) -> None:
+        """Handle removal."""
+        if (dmr_device := self._dmr_device) is not None:
+            self._dmr_device = None
+            dmr_device.on_event = None
+            await dmr_device.async_unsubscribe_services()
+
+        if (upnp_server := self._upnp_server) is not None:
+            self._upnp_server = None
+            await upnp_server.async_stop_server()
+
+    def _on_upnp_event(
+        self, service: UpnpService, state_variables: Sequence[UpnpStateVariable]
+    ) -> None:
+        """State variable(s) changed, let home-assistant know."""
+        force_refresh = False
+
+        self.async_schedule_update_ha_state(force_refresh)
```

Comment:
Maybe just call the upnp part of the update and then call self.async_write_ha_state() here